node/subsystem-util/src/nesting_sender.rs

```diff
@@ -0,0 +1,148 @@
+//! reasonably safe in Rust.
+//!
+//! I very much agree with many points in this blog post for example:
+//!
+//! https://maciej.codes/2022-06-09-local-async.html
+//!
+//! Another very good post by Pierre (Tomaka):
+//!
+//! https://tomaka.medium.com/a-look-back-at-asynchronous-rust-d54d63934a1c
+//!
+//! This module helps with this in part. It does not break the multithreaded by default approach,
+//! but it breaks the `spawn everything` approach. So once you `spawn` you will still be
+//! multithreaded by default, despite that for most tasks we spawn (which just wait for network or some
+//! message to arrive), that is very much pointless and needless overhead. You will just spawn less in
+//! the first place.
+//!
+//! By default your code is single threaded, except when actually needed:
+//!		- need to wait for long running synchronous IO (a threaded runtime is actually useful here)
+//!		- need to wait for some async event (message to arrive)
+//!		- need to do some hefty CPU bound processing (a thread is required here as well)
+//!
+//! and it is not acceptable to block the main task for waiting for the result, because we actually
+//! really have other things to do or at least need to stay responsive just in case.
+//!
+//! With the types and traits in this module you can achieve exactly that: You write modules which
+//! just execute logic and can call into the functions of other modules - yes we are calling normal
+//! functions. For the case a module you are calling into requires an occasional background task,
+//! you provide it with a `NestingSender<M, ChildModuleMessage>` that it can pass to any spawned tasks.
+//!
+//! This way you don't have to spawn a task for each module just for it to be able to handle
+//! asynchronous events. The module relies on the using/enclosing code/module to forward it any
+//! asynchronous messages in a structured way.
+//!
+//! What makes this architecture nice is the separation of concerns - at the top you only have to
+//! provide a sender and dispatch received messages to the root module - it is completely
+//! irrelevant how complex that module is, it might consist of child modules also having the need
+//! to spawn and receive messages, which in turn do the same, still the root logic stays unchanged.
+//! Everything is isolated to the level where it belongs, while we still keep a single task scope
+//! in all non blocking/not CPU intensive parts, which allows us to share data via references for
+//! example.
+//!
+//! Because the wrapping is optional and transparent to the lower modules, each module can also be
+//! used at the top directly without any wrapping, e.g. for standalone use or for testing purposes.
+//!
+//! In the interest of time I refrain from providing a usage example here for now, but would instead
+//! like to point you to the dispute-distribution subsystem which makes use of this architecture.
```

Comment:
nit: I was actually thinking before I got this to this part that an example would be great. 😅 I would like to see one here eventually. We should have a TODO or an issue for later.